### PR TITLE
FCBHDBP-586 creating endpoint to delete days from a plan

### PR DESCRIPTION
**Pull request by marcvictorpassarelli**
# Description

Created the `DELETE /plans/{plan_id}/day` endpoint that receives a list of day ids separated by comma to delete from plan and also delete their respective playlists

## Issue Link

[FCBHDBP-586](https://fullstacklabs.atlassian.net/browse/FCBHDBP-586)

## How Do I QA This

Create a test plan, like the one below with 5 days
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-709bb0c2-98fd-4c84-a47d-3723be3e00b6?action=share&source=copy-link&creator=18435954&ctx=documentation

Now get the plan you just created and notice there are 5 days in the days array. The plan id was saved in a collection variable called `plan_id_delete_day`
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-c9489191-daad-4e9d-8263-1d901dbd8f89?action=share&source=copy-link&creator=18435954&active-environment=810fd94b-9392-43e2-9f13-943e8d323135

With the response from the plan creation it was saved the day ids in collection variables (they have the format dayX_id, like day1_id, day2_id) , now we can pass them (one or more) in this DELETE call separated by comma here
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-1e60d6c1-0762-4e62-a4be-5c38a810a5d5?action=share&source=copy-link&creator=18435954&active-environment=810fd94b-9392-43e2-9f13-943e8d323135

Now do a call to get the plan again and notice that the days you just deleted are not in the days array anymore
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-c9489191-daad-4e9d-8263-1d901dbd8f89?action=share&source=copy-link&creator=18435954&active-environment=810fd94b-9392-43e2-9f13-943e8d323135

You can also try to get the playlist from the day you just deleted and you will noticed you receive a Playlist Not Found
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/request/12519377-f0de3d51-0c10-4b45-bfe2-9f7e38e5dbf1?action=share&source=copy-link&creator=18435954&ctx=documentation

## Screenshots (if appropriate)

[FCBHDBP-586]: https://fullstacklabs.atlassian.net/browse/FCBHDBP-586?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/Http/Controllers/Plan/PlansController.php [hunk 1]
```diff
@@ -669,6 +669,69 @@ public function completeDay(Request $request, $day_id)
...
+        
+        $days = checkParam('days', true);
+        $days_ids = explode(',', $days);
+
+        if(count($days_ids) > 20){
+            return $this->setStatusCode(400)->replyWithError('Too many days provided. Maximum is 20.');
+        }
+
+        if ($days) {
+            $deleted_days = PlanDay::whereIn('id', $days_ids)->where('plan_id', $plan->id);
```
Comment (vichugofsl) [inline app/Http/Controllers/Plan/PlansController.php:725]: @marcvictorpassarelli Let's create a static method to wrap up the entire logic for removing the plan days using the days_id and plan ID. You can check an example in the following and let me give an example:

https://github.com/fullstacklabs/dbp/blob/develop/app/Models/Plan/UserPlan.php#L138

```php
    // LOCATION: dbp/app/Models/Plan/PlanDay.php
    /**
     * Remove the Plan Days records attached to a Plan...........
     *
     * @param int $plan_id
     * @param array $plan_day_ids
     *
     * @return bool
     */
    public static function removePlanDaysByPlanId(int $plan_id, array $plan_day_ids): bool
    {
        try {
            \DB::transaction(function () use ($plan_id, $plan_day_ids) {
                $deleted_days = PlanDay::whereIn('id', $plan_day_ids)->where('plan_id', $plan_id);
                $playlists_ids = $deleted_days->pluck('playlist_id')->unique();
                $playlists = Playlist::whereIn('id', $playlists_ids);
                $deleted_days->delete();
                $playlists->delete();
            });
            return true;
        } catch (\Exception $e) {
            \Log::error('Failed to remove plan days', ['plan_id' => $plan_id, 'plan_day_ids' => $plan_day_ids, 'exception' => $e]);
            return false;
        }
    }
```
Marc, if you can see the above, it is using the concept of statement transactions. It is recommended to use this when you are performing more than one database operation and each operation depends on another. This ensures that if an operation fails, the state of the database can be reverted to the initial state before starting the first operation, avoiding partially completed operations.

You can check the concept of database transactions here: https://laravel.com/docs/11.x/database#database-transactions or https://www.javatpoint.com/mysql-transaction

File: app/Http/Controllers/Plan/PlansController.php [hunk 2]
```diff
@@ -669,6 +669,69 @@ public function completeDay(Request $request, $day_id)
...
+            $playlists_ids = $deleted_days->pluck('playlist_id')->unique();
+            $playlists = Playlist::whereIn('id', $playlists_ids);
+            $deleted_days->delete();
+            $playlists->delete();
+        }
+
+        return $this->reply('Plan days were succesfully deleted');
```
Comment (vichugofsl) [inline app/Http/Controllers/Plan/PlansController.php:732]: @marcvictorpassarelli Here you can validate and return a 500 code if the new method returns false, as the plan days could not be removed, something like:

```php
        if (!PlanDay::removePlanDaysByPlanId(...)) {
            return $this->setStatusCode(SymfonyResponse::HTTP_INTERNAL_SERVER_ERROR)->replyWithError("Plan Days can't be deleted");
        }
        
        return $this->reply('Plan days were succesfully deleted');
```

File: routes/api.php [hunk 3]
```diff
@@ -416,6 +416,9 @@
 Route::name('v4_internal_plans.draft')
     ->middleware('APIToken:check')
     ->post('plans/{plan_id}/draft', 'Plan\PlansController@draft');
+Route::name('v4_internal_plans_days.delete')
```
Comment (vichugofsl) [inline routes/api.php:419]: @marcvictorpassarelli As you have added a new endpoint, we need to run the following command to update the API documentation. After you run the command, you should check if the file `dbp/public/openapi.json` has been updated.

```shell
php artisan openApi:generate
```

**Review by vichugofsl — CHANGES_REQUESTED** [review]
See the above comments

File: app/Http/Controllers/Plan/PlansController.php [hunk 4]
```diff
@@ -669,6 +669,69 @@ public function completeDay(Request $request, $day_id)
...
+     */
+
+    public function deleteDays(Request $request, $plan_id)
+    {
+        // Validate Project / User Connection
+        $user = $request->user();
+        $user_is_member = $this->compareProjects($user->id, $this->key);
+
+        if (!$user_is_member) {
+            return $this->setStatusCode(401)->replyWithError(trans('api.projects_users_not_connected'));
```
[the remark this answers is not in the document]
Comment (marcvictorpassarelli) [reply app/Http/Controllers/Plan/PlansController.php:702]: Noted. I changed the hard coded numbers for the constants

**Review by vichugofsl — APPROVED** [review]

**Review by vichugofsl — APPROVED** [review]
LGTM

**Pull request merged**